src/component/1d/matrix/useMatrix.ts

```diff
@@ -0,0 +1,153 @@
+import { NumberArray } from 'cheminfo-types';
+import { xFindClosestIndex } from 'ml-spectra-processing';
+import { Spectrum } from 'nmr-load-save';
+import { useMemo } from 'react';
+
+import { isSpectrum1D } from '../../../data/data1d/Spectrum1D';
+import { useChartData } from '../../context/ChartContext';
+import useSpectraByActiveNucleus from '../../hooks/useSpectraPerNucleus';
+
+interface MatrixColor {
+  color: string;
+  start: number;
+  end: number;
+}
+
+export function findXFromToIndex(x, options: { from: number; to: number }) {
+  const { from, to } = options;
+  let fromIndex = xFindClosestIndex(x, from);
+  let toIndex = xFindClosestIndex(x, to);
+
+  if (fromIndex > 0) {
+    for (let i = fromIndex; i >= 0; i--) {
+      if (x[i] <= from) {
+        fromIndex = i;
+        break;
+      }
+    }
+  }
+  if (toIndex < x.length) {
+    for (let i = toIndex; i < x.length; i++) {
+      if (x[i] >= to) {
+        toIndex = i;
+        break;
+      }
+    }
+  }
+  return { fromIndex, toIndex };
+}
+
+export function sliceArray(
```

Comment:
It is not clear to me why this method exists and we need to keep the first and last point. Should we not just keep the point just before and after (if they exists !) ? If really required it should be renamed to something like 'sliceArrayAndKeepFirstLast'

Also at the implementation point of view doing an Array.slice will likely be much faster. At worse you could slice 2 points more and change the first and the last one.

And concerning types you call this method with color (an array of string) and NumberArray includes already number[] and Float64Array.